### PR TITLE
Add test generator

### PR DESCRIPTION
**Pull request by senekor**
This adds a test generator written in Rust. The main goal is to be able to quickly synchronize with problem-specifications. In the process, I rewrote a bunch of Bash scripts in Rust to clean up the tooling.

I changed a couple of things in the track config.json, including on deprecated exercises. Mostly for more consistent de-/serialization. Maybe that's a big no-no?

This is the first step - to have a generator in the repo. The next steps will be to apply it one-by-one to all exercises present in the problem-specifications repository. That will probably lead to iterative changes to the generator as well. Once that's done, I will add CI tests to make sure all exercises are in sync with the version of problem-specifications pinned here as a submodule. Updating will then be a matter of pulling the submodule and pleasing CI tests.

The following is intended as guidance for reviews as well as the commit message of the squash commit if this gets merged.

---

Add test generator

* cleanup dev tooling
  * move all scripts into `bin/`
  * turn several scripts into Rust tests:
    * `check_exercises_for_authors.sh`
    * `count_ignores.sh`
    * `verify_exercise_difficulty.sh`
    * `lint_tool_file_names.sh`
    * `lint_trailing_spaces.sh`
  * remove several scripts:
    * `check_uuids.sh`:
      `configlet lint` covers that already
    * `ensure_lib_src_rs_exists.sh`:
      this is not an issue if exercise stubs are generated
    * `clean_topics_vs_practices.py`:
      applies undocumented standards (max 10 practice exercises per topic)
    * `fetch_canonical_data`:
      we now have the submodule for this

* cleanup documentation
  * remove generic instructions about WSL
  * merge `CONTRIBUTING.md` and `maintaining.md`
  * move contributing docs in `README.md` to `CONTRIBUTING.md`
  * outsource instructions for making good PRs
  * outsource instructions for installing Rust
  * steal beautiful `REAMDE.md` from Python track

* apply conventions and guidelines in a couple places
  where the new Rust tests found deficiencies

* replace Bash-based exercise generation with Rust-based one
  (deleting the related `util/` Rust crates)
  * can be used to sync with problem-specifications too
  * uses the Tera templating engine for customizable test generation
    inspired by the python track using Jinja for this purpose

* apply test generation to acronym exercise to validate MVP

### CODE REVIEW

**Comment by senekor** [comment]
After finishing my work less than 24 hours ago, I am surprised to discover that I basically wrote something that had already existed until a couple of months ago.

- [PR that deleted the exercise generator written in Rust](https://github.com/exercism/rust/pull/1635)
- [comment that suggested the deletion](https://github.com/exercism/rust/pull/1635#pullrequestreview-1325866788)
- [commit that deleted the exercise generator](https://github.com/exercism/rust/pull/1635/commits/f8848f2f7e447455a8ccdf3304a81b487bfbfb01)

This had *a lot* of what I have rewritten here. Statically typed de-/serialization of canonical data... heck, it even used the same templating engine as my new generator.

I cannot find a reason documented in that PR for deleting such a huge chunk of work. Can any of the participants give me some more context? @dem4ron @ErikSchierboom @petertseng

**Comment by iHiD** [comment]
I think the story here is:
- Aron wanted to add more exercises and wanted to use a test generator to do so
- Aron was told the old generator was broken and unmaintained, and Aron didn't have a lot of time so he quickly put together some bash scripts that worked well enough.
- During the PR review process it was suggested that the old unmaintained one was removed as part of this
- Remo comes along, sees it would be nice to have a Rust one, and creates a new one.

I don't think Aron's was ever really intended to replace the existing one as much as be something useful that got somewhere quickly. That "replacement" sort of inadvertently happened, and now we've taken a big circle to have a fully working one written in Rust.

From an org perspective, this is one of the painful things about long-term OSS projects where contributors come and move on and work gets abandoned, or redone. We should probably have done a better job of documenting that there was a legacy (or at broken) Rust version of the generator, so that if someone else came along to write one, then they'd see that was there. Ideally, Aron would have seen https://forum.exercism.org/t/advice-for-writing-a-test-generator/7178 and realised, but he didn't see that issue (maybe it would have been wise to tag @dem4ron as you were replacing his work - maybe some idea in future), but that's not a process, just luck really.

So I think that's where we are. Unfortunate series of events that means @senekor has had to duplicate existing work, which I know will be frustrating. But on the plus side, we have a shiny new generator, with an active maintainer, and hopefully we can use it to get the Rust track to a great shape 🙂

**Comment by senekor** [comment]
> Aron was told the old generator was broken and unmaintained

Alright, I guess that's the piece of information that I was missing. @iHiD thanks for the explanation. I'm not too stressed about the extra work I had to do. Greenfield projects are always more fun than maintaining legacy code. As you said, the more interesting question is how we can carry value into the future without starting from scratch every time there's a rotation in maintainership. I guess a big chunk of the problem was the long gap in maintainership. Which is one of the unfortunate things that can happen in OSS.

**Comment by dem4ron** [comment]
Hi Remo,

Firstly, what you've created looks really good.

I sincerely apologize for not documenting the removal of the old tooling. I never imagined it might inconvenience someone down the road, particularly as it was outdated and wasn't being actively maintained. 

The tiny generator script I developed wasn't intended to replace it; I removed the old tooling based on a comment advising me to do so.
In hindsight, delving deeper into the old tooling and focusing my efforts there could have been more productive. However, I began without any grand intentions, simply hoping to quickly add a few exercises.

I regret missing your forum post, where providing this background could have been of great help to you.

**Comment by senekor** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1722** <a href="https://app.graphite.dev/github/pr/exercism/rust/1722" target="_blank">Graphite</a>  👈
    * **PR #1723** <a href="https://app.graphite.dev/github/pr/exercism/rust/1723" target="_blank">Graphite</a> 
      * **PR #1727** <a href="https://app.graphite.dev/github/pr/exercism/rust/1727" target="_blank">Graphite</a> 
        * **PR #1726** <a href="https://app.graphite.dev/github/pr/exercism/rust/1726" target="_blank">Graphite</a> 
        * **PR #1725** <a href="https://app.graphite.dev/github/pr/exercism/rust/1725" target="_blank">Graphite</a> 
        * **PR #1724** <a href="https://app.graphite.dev/github/pr/exercism/rust/1724" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/exercism/rust/1722?utm_source=stack-comment).

File: config.json [hunk 1]
```diff
@@ -1514,23 +1514,26 @@
...
       },
       {
         "slug": "kindergarten-garden",
         "name": "Kindergarten Garden",
         "uuid": "c27e4878-28a4-4637-bde2-2af681a7ff0d",
         "practices": [],
         "prerequisites": [],
-        "difficulty": 1
+        "difficulty": 1,
+        "topics": []
```
Comment (ErikSchierboom) [inline config.json:1527]: Why is the `topics` key added? It is only really there for inspiration for the `practices`/`prerequisites` fields, but it serves no purpose from Exercism's viewpoint.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -4,56 +4,19 @@ name: CI
...
-
-      - name: Check exercises for authors
-        run: ./_test/check_exercises_for_authors.sh
-
-      - name: Ensure relevant files do not have trailing whitespace
-        run: ./bin/lint_trailing_spaces.sh
-
   configlet:
     name: configlet lint
     runs-on: ubuntu-latest
```
Comment (ErikSchierboom) [inline .github/workflows/tests.yml:17]: Something for a separate PR: pin the runner to a specific version

**Review by ErikSchierboom — APPROVED** [review]

File: config.json (hunk 1 above)
Comment (senekor) [reply config.json:1527]: I opened a separate [issue](https://github.com/exercism/rust/issues/1735) for me to consider what to do with this field / how to document it properly. In this PR, they are only added for simple and consistent de-/serialization.

I may or may not clean up empty topic fields when I decide what to do on the separate issue.

**Pull request merged**